### DISCUSSION TITLE
Imitating Named Supertype/Union Class

### DISCUSSION THREAD

**Discussion by RespiteSage** [json-schema-org/community#166] [Q&A] [answered]
I'm working off of a JSON Schema that I and one of my coworkers are using for interprocess communication between programs in two different languages (Java and C++). We're autogenerating code using different tools (as my coworker's tool is Java-specific), and I recently realized that they were using the `extends` syntax, which has apparently been deprecated for quite a long time, to generate code for a superclass without any properties. Here's the relevant section of subschemas (from a file with definitions for a bunch of subschemas that are used in several other schemas):

```json
{
  "distribution": {
    "title": "Distribution",
    "description": "A Probability Distribution",
    "type": "object"
  },
  "normalDistribution": {
    "title": "NormalDistribution",
    "description": "A Normal Probability Distribution",
    "type": "object",
    "extends": {
      "$ref": "#/distribution"
    },
    "additionalProperties": false,
    "required": [
      "mean",
      "standardDeviation"
    ],
    "properties": {
      "mean": {
        "description": "The mean (middle) of the distribution",
        "type": "number"
      },
      "standardDeviation": {
        "description": "The standard deviation (spread) of the distribution",
        "type": "number"
      }
    }
  },
  "triangularDistribution": {
    "title": "TriangularDistribution",
    "description": "A Triangular Probability Distribution",
    "type": "object",
    "extends": {
      "$ref": "#/distribution"
    },
    "additionalProperties": false,
    "required": [
      "maximum",
      "minimum",
      "peak",
      "ratio"
    ],
    "properties": {
      "maximum": {
        "description": "The maximum of the distribution",
        "type": "number"
      },
      "minimum": {
        "description": "The minimum of the distribution",
        "type": "number"
      },
      "peak": {
        "description": "The peak of the distribution",
        "type": "number"
      },
      "ratio": {
        "description": "The ratio of the distribution",
        "type": "number"
      }
    }
  },
  "uniformDistribution": {
    "title": "UniformDistribution",
    "description": "A Uniform Probability Distribution",
    "type": "object",
    "extends": {
      "$ref": "#/distribution"
    },
    "additionalProperties": false,
    "required": [
      "maximum",
      "minimum"
    ],
    "properties": {
      "maximum": {
        "description": "The maximum of the distribution",
        "type": "number"
      },
      "minimum": {
        "description": "The minimum of the distribution",
        "type": "number"
      }
    }
  }
}
```

The issue is that my tool ([quicktype](https://github.com/quicktype/quicktype)) does not implement the `extends` syntax (very understandably), so I'm hoping that there's a way around this other than using `oneOf` with references to all of the "subtype" schemas wherever a `Distribution` is needed. My current approach is to use `oneOf` in the "supertype":

```json
{
  "distribution": {
    "title": "Distribution",
    "description": "A Probability Distribution",
    "type": "object",
    "oneOf": [
      {
        "$ref": "#/normalDistribution"
      },
      {
        "$ref": "#/triangularDistribution"
      },
      {
        "$ref": "#/uniformDistribution"
      }
    ],
    "additionalProperties": false
  },
  "normalDistribution": {
    "title": "NormalDistribution",
    "description": "A Normal Probability Distribution",
    "type": "object",
    "additionalProperties": false,
    "required": [
      "mean",
      "standardDeviation"
    ],
    "properties": {
      "mean": {
        "description": "The mean (middle) of the distribution",
        "type": "number"
      },
      "standardDeviation": {
        "description": "The standard deviation (spread) of the distribution",
        "type": "number"
      }
    }
  },
  "triangularDistribution": {
    "title": "TriangularDistribution",
    "description": "A Triangular Probability Distribution",
    "type": "object",
    "additionalProperties": false,
    "required": [
      "maximum",
      "minimum",
      "peak",
      "ratio"
    ],
    "properties": {
      "maximum": {
        "description": "The maximum of the distribution",
        "type": "number"
      },
      "minimum": {
        "description": "The minimum of the distribution",
        "type": "number"
      },
      "peak": {
        "description": "The peak of the distribution",
        "type": "number"
      },
      "ratio": {
        "description": "The ratio of the distribution",
        "type": "number"
      }
    }
  },
  "uniformDistribution": {
    "title": "UniformDistribution",
    "description": "A Uniform Probability Distribution",
    "type": "object",
    "additionalProperties": false,
    "required": [
      "maximum",
      "minimum"
    ],
    "properties": {
      "maximum": {
        "description": "The maximum of the distribution",
        "type": "number"
      },
      "minimum": {
        "description": "The minimum of the distribution",
        "type": "number"
      }
    }
  }
}
```

The code generated in C++ for this isn't particularly satisfactory, but I don't know whether this is an issue with my schema or something I could request from or contribute to the tool I'm using. Is there a better way to get what I want here?

Notably, this question differs from all the JSON schema "inheritance" questions I could find online because I don't want a supertype with properties. I essentially just want a *named* union subschema.

**Comment by gregsdennis** [comment]
> I recently realized that they were using the `extends` syntax, which has apparently been deprecated for quite a long time...

`extends` was never part of the spec. There have been discussions around it, and at least one implementation supported it (not sure which), but it's not something that was ever official.

I realize this isn't really an answer, but you may be interested in https://github.com/json-schema-org/vocab-idl, which is exploring cases exactly like this in an effort to build a vocabulary so that it _can_ be officially supported.

**Reply by RespiteSage** [reply]
It seems that the tool my coworker is using for generating Java from the schema is [jsonschema2pojo](https://www.jsonschema2pojo.org/). Apparently that tool specifically uses `extends`.

**Reply by awwright** [reply]
Actually "extends" existed for a while, but was removed after [draft-03](https://datatracker.ietf.org/doc/html/draft-zyp-json-schema-03#section-5.26). I forget the exact reason... I think it became obsolete after $ref, and it wasn't implemented consistently.

**Comment by handrews** [comment, marked as answer]
I think one of the reasons that `extends` was dropped gets back to the nature of JSON Schema as a constraint system for validation rather than a data definition system.

Using draft-03's `extends` the way you show here did ***not*** mean that the extended schema (`#/definitions/distribution`) would function like the version that uses `oneOf`.  With that `oneOf` approach `#/definitions/distribution` will validate instances that look like any of those three other schemas.**

However, with `extends`, `#/definitions/distribution` will validate anything and everything, because it has no local constraints, and it is ***unware of the `extends` in the other schema objects.***  Of course, people could implement tools that behaved differently, but that's not actually in the spec (this might have been part of the "wasn't consistently implemented" thing that Austin mentioned, idk, I didn't start using JSON Schema until draft-04).

JSON Schema processing generally works in a parent-to-child fashion, with "child" encompassing `$ref` targets at runtime.  So you can incorporate the constraints of a child schema by `$ref`-ing it from the parent in a `oneOf`, but you can't magically have its effects incorporated by `$ref`-ing the "parent" from the "child".  I put those in quotes because what really happens then is that your "parent" is treated as a child in terms of JSON Schema processing.  So `extends` was always backwards from how the processing actually works.  At least as best I can tell, years and years after `extends` was dropped.

_**except actually it won't work at all, because `"additionalProperties": false` next to a `oneOf` doesn't do what you want it to do.  You want `"unevaluatedProperties": false`, which wasn't introduced until draft 2019-09._
